### ISSUE TITLE
Autocomplete file names adds a slash at the end of the file

### ISSUE THREAD

**Issue by eblis** [chrisant996/clink#329]
I suspect this issue might actually be coming from [clink completions](https://github.com/vladimir-kotikov/clink-completions) instead of clink itself, but I'm not sure how to check.

When I try to autocomplete files for git commands, for example `git add flexpro` and then hit `tab` I get this autocompleted 
![image](https://user-images.githubusercontent.com/540188/179709481-78cdaf5c-5757-4c57-a9a4-debb33b08214.png)
The problem is that this command, if run, won't actually do anything because of the trailing `\`.

How can I prevent this from happening ?
I suspect it's coming from the `clink completions` module because if I type in `ll fle` and then hit `tab` it doesn't add the trailing `'\`.

**Comment by chrisant996** [comment]
I don't get that behavior from Vladimir's [clink-completions](https://github.com/vladimir-kotikov/clink-completions), nor from my fork of [clink-completions](https://github.com/chrisant996/clink-completions) (which has some significant improvements; there is a pull request outstanding back to Vladimir's but there has been no response).

Some troubleshooting questions:
- Can you share the output from `clink set`?
- What other scripts do you have loaded besides clink-completions, clink-flex-prompt, and clink-gizmos?
- Just in case -- is there a directory named "flexprompt_modules.lua"?  Because that would cause that behavior.
- Do all files get an extra slash, or is it only that one specific file?

Also, you could try renaming the `git.lua` file to `git.lua.txt` and start a new session of Clink, and see if the problem still occurs.  If yes, then it's more likely (but not conclusive) that it has to do with clink-completions somehow.  If no, then it's more likely it doesn't have to do with clink-completions (but still not conclusive).

**Comment by eblis** [comment]
* `Clink set` output:
```
autosuggest.async                 True
autosuggest.enable                True
autosuggest.original_case         True
autosuggest.strategy              match_prev_cmd history completion
clink.autostart
clink.autoupdate                  True
clink.colorize_input              True
clink.default_bindings            windows
clink.gui_popups                  False
clink.logo                        full
clink.paste_crlf                  space
clink.path
clink.promptfilter                True
clink.update_interval             5
cmd.admin_title_prefix
cmd.altf4_exits                   True
cmd.auto_answer                   off
cmd.ctrld_exits                   True
cmd.get_errorlevel                True
color.arg                         bold
color.arginfo                     sgr 38;5;172
color.argmatcher                  sgr 1;38;5;40
color.cmd                         sgr 1;38;5;231
color.cmdredir                    sgr 38;5;172
color.cmdsep                      sgr 38;5;214
color.comment_row                 sgr 38;5;87;48;5;18
color.description                 sgr 38;5;39
color.doskey                      sgr 1;38;5;75
color.executable                  sgr 1;38;5;33
color.filtered                    sgr 38;5;231
color.flag                        sgr 38;5;117
color.hidden                      sgr 38;5;160
color.horizscroll                 sgr 38;5;16;48;5;30
color.input                       sgr 38;5;222
color.interact                    bold
color.message                     default
color.modmark
color.popup
color.popup_desc
color.prompt                      red
color.readonly                    sgr 38;5;28
color.selected_completion         sgr 38;5;16;48;5;254
color.selection                   sgr 38;5;16;48;5;179
color.suggestion                  sgr 38;5;239
color.unexpected                  default
color.unrecognized                sgr 38;5;203
debug.log_terminal                False
directories.dupe_mode             add
doskey.enhanced                   True
exec.aliases                      True
exec.commands                     True
exec.cwd                          True
exec.dirs                         True
exec.enable                       True
exec.files                        False
exec.path                         True
exec.space_prefix                 True
files.hidden                      True
files.system                      False
history.dont_add_to_history_cmds  exit history
history.dupe_mode                 erase_prev
history.expand_mode               not_quoted
history.ignore_space              False
history.max_lines                 10000
history.save                      True
history.shared                    False
history.sticky_search             False
history.time_format               %F %T
history.time_stamp                off
lua.break_on_error                False
lua.break_on_traceback            False
lua.debug                         False
lua.path
lua.reload_scripts                False
lua.strict                        True
lua.traceback_on_error            False
match.expand_envvars              True
match.fit_columns                 True
match.ignore_accent               True
match.ignore_case                 relaxed
match.max_fitted_matches          0
match.max_rows                    0
match.preview_rows                5
match.sort_dirs                   with
match.substring                   True
match.translate_slashes           system
match.wild                        True
prompt.async                      True
prompt.transient                  off
readline.hide_stderr              False
terminal.adjust_cursor_style      True
terminal.differentiate_keys       False
terminal.east_asian_ambiguous     auto
terminal.emulation                auto
terminal.mouse_input              auto
terminal.mouse_modifier
terminal.raw_esc                  False
terminal.use_altgr_substitute     True
```
* I'm only using clink-flexprompt and clink-completions, i'm not using clink-gizmos.
* There's no folder in that repo except `.git`.
```
 Directory of D:\work\personal\clink-flex-prompt

18/07/2022  09:44    <DIR>          .
18/07/2022  09:44    <DIR>          ..
18/07/2022  08:48                84 .gitignore
18/07/2022  08:48            35,450 demo.png
18/07/2022  08:52            68,516 flexprompt.lua
18/07/2022  21:41            58,579 flexprompt_modules.lua
18/07/2022  08:48            35,236 flexprompt_wizard.lua
18/07/2022  08:48             1,115 LICENSE
18/07/2022  08:48               341 make_release.cmd
18/07/2022  08:48            24,584 README.md
18/07/2022  08:48    <DIR>          tests
18/07/2022  08:48               238 TODO.md
               9 File(s)        224,143 bytes
               3 Dir(s)  249,669,083,136 bytes free
```
* All the files get that `\` at the end, but not all of the time 🤦‍♂️ 
I haven't been able to figure out when it stops adding the `\`, most of the time it adds it.

I renamed `git.lua` to `git.lua.disabled` and just reloaded scripts with ctrl+x + ctrl+r and it stopped adding the `\` for any files.
As soon as I load `git.lua` it starts happening again.
I also tried your version of `clink-completions` and I get the same behaviour.

**Comment by eblis** [comment]
I'm using git version `2.37.1` if that matters.
I've been playing a bit more with autosuggestions and I can pretty reliably get the issue to occur (or not) by doing the following things:
* type `git add ` and then hit tab
* The suggestions will have `\`, for both files and folders
* Now I delete part or the whole suggestion, but without deleting the space after `git add  ` (this text is here so formatting won't remove the space after add)
* If I press tab now, the autosuggestions no longer contains the `\`.
  * But if I press ctrl + space then the autosuggestion **will still** have the `\` (if it's just one option, otherwise see below)
* If however I delete the suggestion and the space as well (or one of the letters from `add` and then type them in again) if I press tab to autosuggest then i have the `\` again

If I type in `git add ` and then use ctrl+space instead of tab (at this point tab would add the `\`) I get all the suggestions and they all contain the `\`, but when I select them they don't have the `\` in them.
I can navigate through all suggestions and only the folders will have `\` when filled in, but all options show the `\`
![image](https://user-images.githubusercontent.com/540188/179911066-94efa17c-055c-4ac1-b567-df788fb44dca.png)
The same thing happens if I type part of the file name and I hit ctrl+space and there's more than one option
![image](https://user-images.githubusercontent.com/540188/179911474-4692c283-a7fb-4921-a0c1-fc7be1f6064d.png)

If however there's only one option then i get the `\` at the end, even with ctrl+space.

**Comment by eblis** [comment]
So the problem seem to be happening for commands that list `matchers.files` as their input, like `git add` and `git annotate` do.
![image](https://user-images.githubusercontent.com/540188/179914028-7e61b8de-e785-48ad-939e-a56da30d53a3.png)
I have modified the `matchers.lua` file to print all matches before returning and this is the output:
![image](https://user-images.githubusercontent.com/540188/179914282-285c73aa-a05e-425b-b225-c2fa4e6d3532.png)
The files as returned by the matchers.files do not contain the `\`, so the problem occurs later somehow ?
But if I understand the code correctly afterwards it's just passed to `clink.arg.new_parser`. Could that be causing this ?

Like I mentioned previously `git checkout` which uses `local function checkout_spec_generator(token)` doesn't experience the same problem.

**Comment by eblis** [comment]
I've played some more with the code in `matchers.lua` and I can get the issue to stop reproducing, but I'm not sure about the side effects of my changes.
With this version of the code (unchanged)
![image](https://user-images.githubusercontent.com/540188/179920408-8332c6b9-abd6-4ddf-a1b1-2003934b1ce5.png)
I get the file name suggestions with `\` at the end, like this:
![image](https://user-images.githubusercontent.com/540188/179920305-66ad49d3-1682-4b7f-bbe4-a76409334ba4.png)

If however I change the call to this (which matches the call that `checkout_spec_generator(token)` does):
![image](https://user-images.githubusercontent.com/540188/179920585-e36e29f6-1fef-4b64-8aa6-5e05cdc03645.png)

Then I no longer get the `\` at the end, neither for files nor folders (`modules` is the modules folder for `clink-completions`)
![image](https://user-images.githubusercontent.com/540188/179920996-f353eb73-e176-4430-8b5f-50a8a706e352.png)
![image](https://user-images.githubusercontent.com/540188/179921047-53a2b68e-f68b-4494-a3a9-9ecaa5836d7b.png)

**Comment by eblis** [comment]
I deleted the whole click-completions folder and re-added everything from your repo, I have the same symptoms.

And I just figured out why ..
I was using your `master` branch, didn't notice you created a new branch called `clink_improvements`. If I use this branch then I can't reproduce the problem anymore

**Comment by chrisant996** [comment]
This fix is now available in v1.3.37.
